lib/identity_cache/query_api.rb

```diff
@@ -14,9 +14,30 @@ def exists_with_identity_cache?(id)
         !!fetch_by_id(id)
       end
 
-      # Default fetcher added to the model on inclusion, it behaves like
-      # ActiveRecord::Base.where(id: id).first
-      def fetch_by_id(id, includes: nil)
+      # Fetch the record by its primary key from the cache or read from
+      # the database and fill the cache on a cache miss. This behaves like
+      # `where(id: id).readonly.first` being called on the model.
+      #
+      # @param id Primary key value for the record to fetch.
+      # @param includes [Hash|Array|Symbol] Cached associations to prefetch from
+      #   the cache on the returned record
+      # @param fill_lock_duration [Float] If provided, take a fill lock around cache fills
+      #   and wait for this duration for cache to be filled when reading a lock provided
+      #   by another client. Defaults to not setting the fill lock and trying to fill the
+      #   cache from the database regardless of the presence of another client's fill lock.
+      #   Set this to just above the typical amount of time it takes to do a cache fill.
+      # @param lock_wait_limit [Integer] Only applicable if fill_lock_duration is provided,
+      #   in which case it specifies the number of times to do a lock wait. After the first
+      #   lock wait it will try to take the lock, so will only do following lock waits due
+      #   to another client taking the lock first. If another lock wait would be needed after
+      #   reaching the limit, then a `LockWaitTimeout` exception is raised. Default is 2. Use
+      #   this to control the maximum total lock wait duration
+      #   (`lock_wait_limit * fill_lock_duration`).
```

Comment:
The API here is awkward and the abstraction is leaky. From a user's perspective, what we care about is "how long do I want to wait", and we should really specify that directly. We want to specify a minimum, based on expected refill time, and a maximum we're willing to tolerate. The `lock_wait_limit` leaks an implementation detail that users shouldn't have to care about (and which complicates tweaking the algorithm later).

Reply:
Specifying a number of times to retry doesn't seem like a leaky abstraction and seems like it is how we normally handle retries.  In fact, I can't think of a place where we do retries for a certain amount of time instead of having a retry limit.

As you mentioned, we need to know how long to wait based on the expected refill time, so I it seems like `fill_lock_duration` is necessary.

What type of tweaks did you have in mind that the current configuration doesn't provide a suitable abstraction for?  It seems to me that we could use the same configuration if we switched to redis to store the lock or if we just did it in-memory in a multi-threaded process.

Reply:
Exponential backoff is a pretty common approach that would map awkwardly to this API.

My point about leaky abstractions is that users want to specify their desired constraints under contention. It's awkward to expose the knobs as "how often should I try" and "how many times should I try" when the user really wants to tell us "I expect the refill to take X seconds" and "I don't want to wait longer than Y seconds".

Reply:
What you are proposing is just a different abstraction.  That doesn't make the configuration I proposed a leaky abstraction, it just isn't an abstraction around the change you want to be able to make.

As a user of the API, I would hate for it to silently change to exponential backoff without explicitly opting into that.  Since the sleep is synchronous, we are wasting the threads time which could eat into capacity.  We don't want to wait for a problem to be resolved in the current process, we want to fail fast if we know there is a problem and rely on semian to backoff in our attempts to retry accessing the resource.

Also, as a user of the API, I find the configuration API you are describing awkward to express what I actually want.  Since this is specifically to solve the thundering herd problem, I really do care about how long the fill lock is taken, since that limits the frequency of attempts to fill the cache key under contention.  I want it to do at most 2 lock waits because that both gracefully handles cache fills that take up to double the typically allowed time and because it allows it to be interrupted once by a cache invalidation.  If I try to express the same thing as a a duration (i.e. `lock_wait_limit * fill_lock_duration`) then it is no longer precise due to floating point math.  Depending on the behaviour chosen for the remaining time, it could either lose the second attempt if there is just under the fill lock duration remaining or perform a last attempt immediately after the second last attempt if there is just over the fill lock duration remaining.

I think if we build something that works well for us, then it is more likely to be suitable for other users that encounter the same thundering herd problem.